File: 08-Greedy-Algorithm/cpp/03-job-sequencing-problem.cpp
class Solution {
public:
    vector<int> JobScheduling(vector<vector<int>>& Jobs) {
        sort(Jobs.begin(), Jobs.end(), [](const vector<int>& a, const vector<int>& b) {
            return a[2] > b[2];
        });

        int maxDeadline = 0;
        for (const auto& job : Jobs) {
            maxDeadline = max(maxDeadline, job[1]);
        }

        vector<int> slot(maxDeadline + 1, -1);
        int countJobs = 0, totalProfit = 0;

        for (int i = 0; i < (int)Jobs.size(); i++) {
            int deadline = Jobs[i][1];
            int profit = Jobs[i][2];

            for (int j = deadline; j > 0; j--) {
                if (slot[j] == -1) {
                    slot[j] = i;
                    countJobs++;
                    totalProfit += profit;
                    break;
                }
            }
        }

        return {countJobs, totalProfit};
    }
};

File: 08-Greedy-Algorithm/python/03_job_sequencing_problem.py
class Solution:
    def JobScheduling(self, Jobs: List[List[int]]) -> List[int]:
        Jobs.sort(key=lambda x: x[2], reverse=True)

        max_deadline = max(job[1] for job in Jobs)
        slot = [-1] * (max_deadline + 1)

        count_jobs = 0
        total_profit = 0

        for i, job in enumerate(Jobs):
            deadline, profit = job[1], job[2]
            for j in range(deadline, 0, -1):
                if slot[j] == -1:
                    slot[j] = i
                    count_jobs += 1
                    total_profit += profit
                    break

        return [count_jobs, total_profit]
